### PR TITLE
Upstream new boto_s3 execution/state modules

### PR DESCRIPTION
**Pull request by aneeshusa**
### What does this PR do?
Add a new `boto_s3` execution and state module,
which currently support managing S3 objects.

We are currently using these internally, however I now see Salt already has an `s3` execution module. I'd prefer to lean on boto3 instead of a Salt-internal AWS interface; let me know your opinions on how to best integrate this functionality.

### What issues does this PR fix or reference?
None AFAIK

### Previous Behavior
Could not use boto from a state to upload a file to S3.

### New Behavior
S3 object state functionality is now available.

### Tests written?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

**Review by ryan-lane — APPROVED** [review]
Already reviewed internally. Works well, and is well implemented.

**Comment by tkwilliams** [comment]
I'm wondering about the advisability of having service-wide boto_s3 modules, rather than a host of per-resource modules.  The latter seem to add a lot of overhead, and wasted boilerplate code. Also, there will very likely be some module (and possibly state) functions which don't directly relate to a given resource type and thus wouldn't have a logical place to live.  I'm suggesting we instead call this module boto_s3, and use it to implement all NEW boto3 based s3 functionalities going forward.  Eventually, it would be nice to move what is currently handled by boto_s3_bucket in here as well...

**Review by tkwilliams — CHANGES_REQUESTED** [review]
As mentioned in my thread comment, I'd like to discuss renaming these modules to boto_s3, rather than boto_s3_object, with the intention of using them generically for boto s3 functionality going forward.

**Comment by aneeshusa** [comment]
That sounds good to me, here's what I'm concretely thinking about renaming these functions to:

Exec module functions:
- `boto_s3.upload_object`
- `boto_s3.get_object_metadata`

State module functions:
- `boto_s3.object_present`

**Comment by tkwilliams** [comment]
That sounds perfect :)  Thanks @aneeshusa, this will make extending it later a breeze...

File: salt/modules/boto_s3.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        )
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Message'] == 'Not Found':
+            return {'result': None}
+        return {'error': __utils__['boto3.get_error'](e)}
+
+    return {'result': metadata}
+
+
+def upload_object(
```
Comment (tkwilliams) [inline salt/modules/boto_s3.py]: Your implementation of upload_object() seems good, nice and straightforward.  Wondering if there would be any benefit to using `put_object()` rather than `upload_file()` - the former provides a lot more options when putting the object (setting Tags, StorageClass, Metadata, etc, etc.), while the latter is very easy to implement :)  I can definitely see both sides here.  I would recommend that, if you do elect to use `upload_file()` for this function, that the function itself be called `upload_file()` rather than `upload_object()` since the latter might confuse someone, and also makes referencing the relevant AWS API docs or boto3 pages a bit harder.  In general I think it's good practice to name execution functions, where possible, to match whichever boto3 functions they are calling under the hood.

File: salt/modules/boto_s3.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    .. code-block:: bash
+
+        salt myminion boto_s3.get_object_metadata \\
+                         my_bucket/path/to/object \\
+                         region=us-east-1 \\
+                         key=key \\
+                         keyid=keyid \\
+                         profile=profile \\
+    '''
+    bucket, _, key = name.partition('/')
```
Comment (tkwilliams) [inline salt/modules/boto_s3.py]: Is it intended that this `key` var overwrite any value for `key` given in the function params above?  I don't imagine anyone is using objects named after their AWS access keys :)

Can I suggest maybe something like the following to help simplify the code, and avoid the name clash? 
```
def get_object_metadata(
    name,
    region=None,
    key=None,
    keyid=None,
    profile=None,
    **kwargs
):
    '''
    Get metadata about an S3 object.
    Returns None if the object does not exist.

    CLI Example:

    .. code-block:: bash

        salt myminion boto_s3.get_object_metadata \\
                         my_bucket/path/to/object \\
                         region=us-east-1 \\
                         key=key \\
                         keyid=keyid \\
                         profile=profile \\
    '''
    if 'Bucket' not in kwargs:
        kwargs['Bucket'], _, kwargs['Key'] = name.partition('/')

    conn = _get_conn(region=region, key=key, keyid=keyid, profile=profile)

    try:
        metadata = conn.head_object(**kwargs)
    except botocore.exceptions.ClientError as e:
        if e.response['Error']['Message'] == 'Not Found':
            log.debug('S3 object not found: {0}/{1}'.format(kwargs['Bucket'],
                                                            kwargs['Key']))
            return {'result': None}
        return {'error': __utils__['boto3.get_error'](e)}

    return {'result': metadata}
```

File: salt/states/boto_s3.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+def __virtual__():
+    '''
+    Only load if boto is available.
+    '''
+    if 'boto_s3.get_object_metadata' not in __salt__:
+        return False
+    return 'boto_s3'
+
+
+# Keys for `extra_args` that we support.
```
Comment (tkwilliams) [inline salt/states/boto_s3.py]: If you use the approach I mention above, these arrays become unneeded, as AWS itself will happily fact check all your args and notify if any are malformed, or invalid.  You can still do any merging needed of "extra_args_from_pillar" into the passed in parameters, as you do below.  As a benefit, you'll find that the currently unsupported options you mention magically start working, as long as they're correctly specified by the caller (e.g. the pillar source).

File: salt/modules/boto_s3.py (hunk 2 above)
Comment (aneeshusa) [reply salt/modules/boto_s3.py]: Ah, this is a mistake I didn't catch while refactoring, I meant to name it `s3_key`. Thanks for catching this! I'll fix it.

File: salt/states/boto_s3.py (hunk 3 above)
Comment (aneeshusa) [reply salt/states/boto_s3.py]: I did originally just pass through the args directly, but added this validation after hitting some issues. Unfortunately, not all of the arguments are handled symmetrically between what we send to S3 and what S3 stores/returns. For one, we'll need to send different subsets of args to different S3 APIs (see `UPLOAD_ONLY_ARGS` and `GET_METADATA_EXTRA_ARGS`). Second, some args that are valid for upload aren't returned (`RequestPayer` vs `RequestCharged`), so we need to exclude those from the diff. Additionally, the various `Grant*` options on upload act as shortcuts for various ACL settings, and are not returned as-is but in a converted form in the ACL data.

In short, we can't just pass all args to all the S3 endpoints we hit and rely on AWS validation, and we also need to add custom code that handles conversions properly for the purposes of diffing and comparing changes. It's important to me that we don't spuriously report changes, because it causes us to make unnecessary API calls, hit ratelimits and chew bandwidth.

File: salt/modules/boto_s3.py (hunk 1 above)
Comment (aneeshusa) [reply salt/modules/boto_s3.py]: As far as I understand, both APIs provide the same options, but `upload_file` just bundles them all under the `ExtraArgs` kwarg, so I think that makes the code a bit simpler here.

I'll rename this to `upload_file`, makes sense.

**Review by aneeshusa — COMMENTED** [review]
I appreciate the review.

**Comment by cachedout** [comment]
Hi all. I'm just letting this discussion continue but if I'm needed here please just ask. I'll have a look at the final agreed-upon implementation once all issues have been worked out. Thanks. :]

File: salt/states/boto_s3.py (hunk 3 above)
Comment (tkwilliams) [reply salt/states/boto_s3.py]: Understood, I had to do something similar in the boto3_elasticache module.  MOST of the params mapped directly, but some were renamed while others returned different formats (on describe) than what had been sent (by create or modify).  Some were ONLY valid for modify operations, not creates, etc, etc  It's annoying for sure.

Even so, there were a large number of args, the vast majority of which DID map cleanly and easily, so I took the approach of simply munging the few that needed "touching" into the common format for comparisons to work right.  But that's certainly an implementation choice which can fall either way :)

If you want a view on one way to manage this sort of "munging" where needed but otherwise letting things pass through where possible, take a look at {modules,states}/boto3_elasticache.py, particularly`_diff_cache_cluster()`.

I absolutely understand the desire to avoid unneeded change records being sent - we're not nearly at your scale and we already get plenty of throttles from the API as it is...

TL;DR  I'm fine with the implementation as it stands, and will happily approve the change set, other than the one following point, which I hope you'll take to heart and address if you can.

There is a huge forward-looking benefit to keeping camelcase args in that format, rather than mapping them to arbitrary snake_case params.  To be honest, the latter is just buying yourself a lifetime of mapping woes, most of which (not all as you point out but the vast majority) would be avoided if you leveraged the boto3 parameter names (and function names as mentioned above) wherever possible.  The reason snake_case was the norm in the original boto2 based modules was because boto2 ITSELF had mapped all the camelcase stuff to snake, and thus snake_case was the way to directly map salt params to boto params.

This is no longer the case - boto3 uses CamelCase, and creating lots of unneeded code-lines simply to convert the one to the other because it looks familiar from writing boto2 code is, as mentioned, just buying future you (and future me and every other developer touching it) a lot of unnecessary grief and LOC.

I know it's a pain when the code is working already as-is, but throwing away all my other input and feedback, the strongest suggestion I can make for this and future boto3 based code is to PLEASE use the native CamelCase NOW, so that we don't have to support complex parameter munging for EVERY parameter -- not just those small few which actually need  munging to work right -- in this module FOREVERMORE :)

Comment (aneeshusa) [reply salt/states/boto_s3.py]: I agree that munging names is annoying. This patch actually doesn't munge any of the possible `ExtraArgs` parameters, just use the frozensets to determine which ones to pass to which S3 API calls or the diffing code. The only munging here is that the state name is used to derive the `Bucket` and `Key` args, so that the user can specify a simple `some-bucket/some/object/path` as the state name, for ergonomics and to fit Salt conventions.

Comment (aneeshusa) [reply salt/states/boto_s3.py]: Oh, and the module injects the content hash into the metadata for diffing purposes, but I'm using the native CamelCase `Metadata` key to access that.

File: salt/modules/boto_s3.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        return (False, msg.format(submsg))
+    return True
+
+
+def __init__(opts):  # pylint: disable=unused-argument
+    if HAS_BOTO:
+        __utils__['boto3.assign_funcs'](__name__, 's3')
+
+
+def get_object_metadata(
```
Comment (tkwilliams) [inline salt/modules/boto_s3.py]: I'm so sorry -- I got into a yellowjacket nest in the yard last Friday and have been largely useless since then.  This is my first chance to check back on this.

I'm a bit confused by this function.  You pass `**extra_args` (which is really just `**kwargs`) to the `head_object()` function without any munging or even validation, but you require them to be passed as `extra_args` rather than the seemingly more compact `**kwargs`.  Since you're doing no validation and no munging, I would request you change the function signature to 
```
def head_object(
    name,
    region=None,
    key=None,
    keyid=None,
    profile=None,
    **kwargs
)
```
Renaming to `head_object()` lets the salt exec function name match the underlying function it is directly calling, which makes documentation lookups for the end-user much simpler.  It also allows you to remove the `if extra_args is None:` stanza since kwargs will always be a dict.

I also suggest (request) that `Bucket` and `Key` be parsed from `name` ONLY in the event they are not explicitly passed.  Allowing arbitrary state names -- e.g. `name="Ensure my file in in the bucket"` is a standard salt vernacular.  Also, the behaviour of passing e.g. `Bucket='my_bucket', Key='path/to/file'` in your current code is undefined, since `Bucket` and `Key` will be defined twice in the `head_object()` call as it is written.

File: salt/modules/boto_s3.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        )
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Message'] == 'Not Found':
+            return {'result': None}
+        return {'error': __utils__['boto3.get_error'](e)}
+
+    return {'result': metadata}
+
+
+def upload_file(
```
Comment (tkwilliams) [inline salt/modules/boto_s3.py]: I think if you look at some of the earlier boto2 modules which Ryan and team developed, you'll see an initial approach of using `name` as the exclusive source for the primary state args, followed by a back-step where `name` is split off to allow for state names of the form `Ensure a file is in S3`, and adding explicit params for the state arg(s).  In this concrete case, I would request that `name` be used as a FALL-BACK source for `Bucket` and `Key` if the latter are not explicitly provided but separate params for the latter be added to allow arbitrary (descriptive rather than syntactic) `name` strings.  I gave an example of how to do this earlier in my code snippets.  Those should be CamelCase for ease of use, and I would think that ExtraArgs should also be camelcase, since it IS an explicit param which is sent to `upload_file()`.  That is, given a signature of
```
upload_file(Filename, Bucket, Key, ExtraArgs=None, Callback=None, Config=None)
```
`FileName`, `Bucket`, `Key`, `ExtraArgs` (and `Callback` and `Config` if they're ever added at a later date) should all be mapped directly to their CamelCase names, rather than munged into snake_case.  This holds true whether one is internally munging some subset of the data inside of ExtraArgs or not.  Also, renaming `FileName` to `source` is non-optimal for the same reasons - please leave it as `FileName` for ease of reference.  Ideally, `Callback` and `Config` would be included in the initial function, with doc strings mentioning they are not yet implemented (if simply passing through a valid value is not enough to make them work).

This should be done for both the execution module functions, AND the state module functions, wherever a param is mapped to one in the execution module function being called inside the state function.

So what I'm hoping to see -- if simply using `**kwargs` is not an option -- is all function params for `upload_file()` to be explicitly defined in the execution module function signature and have the same CamelCase param names, and that any args in the state function which map to execution function names do the same.  That is, the function signature in the exec module should be
```
upload_file(name, Filename, Bucket=None, Key=None, ExtraArgs=None, region=None, key=None, keyid=None, profile=None)
```
Then ensure that `Bucket` and `Key` are only parsed from `name` if `Bucket` and `Key` are not provided on the commandline.

The goal of all this is to 1) make the upstream documentation of AWS and boto3 as DIRECTLY APPLICABLE and USEFUL as possible, and 2) make the job of future maintainers easier by reducing the amount of "guaranteed overhead" they have to take on just to add new functionality.  The more we can map things directly, and SIMPLY without lots of "touching", the easier people's live will be later, both end-users and maintainers.

**Review by tkwilliams — CHANGES_REQUESTED** [review]
Please see the latest inline comments below :)

File: salt/modules/boto_s3.py (hunk 4 above)
Comment (aneeshusa) [reply salt/modules/boto_s3.py]: The `extra_args` naming comes directly from the parameter name for upload_file, and the set of possible arguments is documented at http://boto3.readthedocs.io/en/latest/reference/customizations/s3.html?highlight=s3transfer#boto3.s3.transfer.S3Transfer.upload_file.
The bucket and key arguments are separate even in the boto3 API, so I'd prefer to keep them separate from extra args in the Salt API as well.
Salt provides the ability for arbitrary state IDs, but the state names are semantically significant. I agree that arbitrary IDs that are human readable are useful and we use them as well, but the name is meant to refer to the actual resource to be managed - see the examples in the 10-minute Salt tutorial (https://docs.saltstack.com/en/getstarted/config/functions.html).

File: salt/modules/boto_s3.py (hunk 5 above)
Comment (aneeshusa) [reply salt/modules/boto_s3.py]: I'm OK with using CamelCase args for the actual upload options (e.g. `ServerSideEncryption`), but I think following Salt conventions with snake_case for top level arguments looks nicer and is easier to follow.
I've used `source` instead of `FileName` to match the `file` states (e.g. `file.managed`) among others.

I think the benefits of a little sugar here for consistency with Python style and other Salt APIs outweigh those of directly shelling out to boto3. Trying to map all of the possible arguments is futile, but considering there's only ~3 top level arguments right now (name, source, and extra_args), I don't think this adds a significant maintenance burden.

**Review by aneeshusa — COMMENTED** [review]
Thanks for the review, @tkwilliams. I've left details inline, but I think the remaining sugar here is more beneficial than not.
@cachedout, happy to hear your thoughts here as well.

File: salt/modules/boto_s3.py (hunk 4 above)
Comment (tkwilliams) [reply salt/modules/boto_s3.py]: My question above was not about `upload_file()` but rather `head_object()`, which you use more-or-less directly in `get_object_metadat()` - `head_object()` does not have an `extra_args` (or an `ExtraArgs`) parameter, so as I mentioned I am confused as to why it was introduced here (or anywhere actually except in `upload_file()` where it actually is a function param (as `ExtraArgs`)  Again, my concern is about adding unneeded complexity (that is to say, `rakes`) for no particular reason, which will only serve to make maintainer's lives harder later, and yields no visible benefit now to justify it.

File: salt/modules/boto_s3.py (hunk 5 above)
Comment (tkwilliams) [reply salt/modules/boto_s3.py]: The major gist of my requests is, in fact, to MAP things as directly to the boto3 api as possible.  For reasons previously outlined, this makes maintainability, and future extension, as WELL as end-user ease of use, MUCH better.  The REASON snake_case is the norm for the existing boto2 based modules is because that's what boto2 DID, and thus the conventions of the time demanded mapping salt parameter names as directly as possible to the underlying boto2 params.  As previously mentioned, this is no longer the case, so the need is not for blindly following a given pattern when the rationale for it has changed, but rather to update the pattern when needed so the rationale which drives it is still served.

A side benefit is that you don't have to build extensive, and exhaustive, mapping tables of snake_case-to-CamelCase args, and update them every time something changes in boto3.  By utilizing a default pass-through approach, only munging params when AWS or boto force it due to brain-deadness on their part, you reap huge benefits in terms of maintainability and end-user convenience (not to mention the current development effort, which is greatly reduced, to boot).

**Comment by cachedout** [comment]
Hi @aneeshusa and @tkwilliams. Thanks for the great discussion here.

This is starting to look really good to me. @tkwilliams do you have unresolved concerns at this point beyond styling or are we closing in on getting this merged?

**Comment by tkwilliams** [comment]
Hi @cachedout,

Actually, most of my concerns are still unresolved.  The questions, and comments, of 10 days ago haven't been addressed quite yet.

I have a patch I worked on last week, but've been out for vacay all this week, with today my first day back.  I'll polish it up in the next day or so and PR it so my recommendations can be more concrete.

**Comment by tkwilliams** [comment]
Sorry, I really am working on this - just been behind the bus the past couple of weeks, and now Denise has asked me to develop a whole new set of salt modules for ElasticMapReduce by the end of the week :-/  We'll see how THAT goes.

**Comment by ryan-lane** [comment]
I'm in agreement with @aneeshusa here. There's very few top level arguments, and it's nicer for them to match salt's style, rather than for them to be 1:1 with the boto3 api. If this was a large effort to continue mapping things in, I'd feel otherwise, but S3's API is old and very stable and the chances of needing to do a lot of ongoing work to continue mapping things is low.

I think we're roughly at odds in manner of style here, and it would be good to get this in, rather than it languishing. There's not really anything here that seems to be a showstopper.

**Comment by rallytime** [comment]
I agree with @ryan-lane on this. We should get this in for Oxygen instead of letting it get stale since the issues raised here are largely a difference in style. I'd rather we match Salt's styling, as defined [here](https://docs.saltstack.com/en/latest/topics/development/conventions/style.html#variables) and then update things as needed moving forward.
